internal/controller/linodefirewall_controller.go

```diff
@@ -203,6 +205,32 @@ func (r *LinodeFirewallReconciler) reconcile(
 	return ctrl.Result{}, nil
 }
 
+func (r *LinodeFirewallReconciler) removeAddressSetFinalizer(ctx context.Context, logger logr.Logger, fwScope *scope.FirewallScope, addrSetRef *corev1.ObjectReference) error {
+	if addrSetRef == nil || r.Client == nil {
+		return nil
+	}
+	addrSet := &infrav1alpha2.AddressSet{}
+	if addrSetRef.Namespace == "" {
+		addrSetRef.Namespace = fwScope.LinodeFirewall.Namespace
+	}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: addrSetRef.Namespace, Name: addrSetRef.Name}, addrSet); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.Error(err, "failed to fetch referenced AddressSet", "namespace", addrSetRef.Namespace, "name", addrSetRef.Name)
+			return err
+		}
+	}
+	finalizer := fmt.Sprintf("lfw.%s.%s", fwScope.LinodeFirewall.Namespace, fwScope.LinodeFirewall.Name)
+	if controllerutil.ContainsFinalizer(addrSet, finalizer) {
```

Comment:
i don't think we need to check before calling `removeFinalizer` - we can replace this as
```
if controllerutil.RemoveFinalizer()... {
}
```